working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+
+A *mixin application* receives corresponding forwarding constructors for all accessible generative constructors of the superclass. The forwarding constructor has the same basename (identifer after the class name for named constructors, empty for the unnamed constructor), the same parameter signature as the superclass constructor, including default values for optional parameters, and an initializer list with a `super`-invocation to the corresponding superclass constructor which forwards each parameter as an argument to the corresponding superclass constructor parameter. The forwarding constructor is `const` if the superclass constructor is `const` *and* the mixin declares no instance variables.
+
+Since a mixin application class cannot contain static members, there is no risk of the introducing a conflict between a constructor name and a static member. 
+
+### Goal
+
+We combine the forwarding constructor feature with default constructors and automatic initialization of declared instance variables. This *unifies* default constructors and mixin application forwarding constructors *and* adds more power to the default constructor for simple classes.
+
+We expect this to happen after non-nullable types have been added to the language, so the proposal is written in terms of that, with a few guidelines for how legacy libraries work where nullability matters.
+
+## Proposal
+
+Let *C* be a class declaration with name `C` and superclass *S* with name `S` and which *declares* no constructor.
+
+It is a compile-time error if:
+
+- *C* declares a private-named and potentially non-nullable instance variable with no initializer expressions. 
+- Or in **NNBD-legacy mode**,  *C* declares a private-named and final instance variable with no initializer expression.
+
+Otherwise, for each constructor *s* of the superclass, named `S` or `S.id`, satisfying all of the following:
+
+1. If *s* is a named constructor with name `S.id` then:
+   1.  *s* is accessible ( `id` is not a private identifier or *S* is not declared in a different library), and
+   2.  *C* does not declare a static member named *id*.
+2. The *s* constructor is a generative constructor.
+3. *C* does not declare an instance variable with no initializer expression and with the same basename as a required named parameter of *s*.
+4. Either *C* does not declare any instance variables or *s* does not have any optional positional parameters (required because we can't declare a constructor with both optional position parameters and named parameters).
+
+add a generative default constructor *c* to *C* with the corresponding name `C` or `C.id`, the following parameters:
+
+- For each non-private instance variable with no initializer expression declared by *C*, where `v` is the name of the instance variable, the constructor has a named initializing formal parameter `this.v` with no default value. The named parameter is *required* if the instance variable has a potentially non-nullable type.
```

Comment:
This makes it a breaking change to switch between a field and getter unless you remember to also rewrite the constructor.

Reply:
All the more reason to do sealed by default. :)

Reply:
If we can't allow that, then default initializing constructors are impossible.
It's also a breaking change to change a field to a getter/setter with an *explicitly written* constructor, because you can no longer use `this.x` to initialize it.

I think it's reasonable. You are editing a class, and the implicit constructor added to that class may change depending on the editing. You are in the perfect position to fix that too, and your tests should catch the change.